### PR TITLE
use correct alive bytes in ancient pack

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem
We were previously using the total # of alive bytes from all ancient storages that could be shrunk to determine how many storages we need.
This means we think we need far more packed storages than we actually do need. This causes us to skip packing more storages than we should.

#### Summary of Changes
Use only the # of alive bytes that we are actually going to pack to determine how many ideal storages are needed.

Fixes #

### CODE REVIEW

**Comment by jeffwashington** [comment]
a recent mnb validator log:
```
shrink_ancient_stats
...
many_ref_slots_skipped=570
```

We skipped 570 slots with many refs in them. This is not the idea we were after.

**Comment by HaoranYi** [comment]
https://github.com/anza-xyz/agave/blob/4a777d1288ec7d935db8d8d45c3eff16a4ade0a3/accounts-db/src/ancient_append_vecs.rs#L175

Should we recalculate target shrink alive_bytes here too?

**Comment by jeffwashington** [comment]
> https://github.com/anza-xyz/agave/blob/4a777d1288ec7d935db8d8d45c3eff16a4ade0a3/accounts-db/src/ancient_append_vecs.rs#L175
> 
> Should we recalculate target shrink alive_bytes here too?

no. This is what the existing calculation is for, I think. We want to shrink a certain percent of the overall shrinkable alive bytes. I think this is the correct usage.

**Comment by HaoranYi** [comment]
> no. This is what the existing calculation is for, I think. We want to shrink a certain percent of the overall shrinkable alive bytes. I think this is the correct usage.

I see. Yes. we use that to turn off the should_shrink flag, which is the reason why we need to recalculate the actual size. Thanks for the explaination.

File: accounts-db/src/ancient_append_vecs.rs [hunk 1]
```diff
@@ -739,11 +738,16 @@ impl AccountsDb {
...
         tuning: &PackedAncientStorageTuning,
-        alive_bytes: u64,
         mut many_ref_slots: IncludeManyRefSlots,
     ) -> AccountsToCombine<'a> {
+        let alive_bytes = accounts_per_storage
+            .iter()
+            .map(|a| a.0.alive_bytes)
+            .sum::<u64>();
+
+        log::error!("alive bytes: {}", alive_bytes);
```
Comment (HaoranYi) [inline accounts-db/src/ancient_append_vecs.rs]: nit: do we need this log?

**Review by HaoranYi — APPROVED** [review]
lgtm!

**Comment by HaoranYi** [comment]
In the following code, we are using `min_resulting_packed_slots` to skip ancient slots that have many_refs_newest_alive in packing. 

min_resulting_packed_slot is calculated from the alive_bytes/storage_size. With this PR, we get a better bound of alive_bytes and reduce the many_ref_slots_skipped! 

But we can do better, I think. 

Since we know that many_refs_newest_alive will be packed to highest slots first before 1-ref accounts. 

We just need to skip the number of maximum slots needed to pack all many_ref_newest_alive, which can be calculated with the sum of bytes for many_ref_newest_alive/ideal_storage_size. This will give us an even smaller many_ref_slots_skipped. 

Does this sound reasonable? 

```
if many_ref_slots == IncludeManyRefSlots::Skip
                && !shrink_collect
                    .alive_accounts
                    .many_refs_this_is_newest_alive
                    .accounts
                    .is_empty()
            {
                let mut required_packed_slots = min_resulting_packed_slots;
                if many_refs_old_alive.accounts.is_empty() {
                    // if THIS slot can be used as a target slot, then even if we have multi refs
                    // this is ok.
                    required_packed_slots = required_packed_slots.saturating_sub(1);
                }

                if (target_slots_sorted.len() as u64) >= required_packed_slots {
                    // we have prepared to pack enough normal target slots, that form now on we can safely pack
                    // any 'many ref' slots.
                    many_ref_slots = IncludeManyRefSlots::Include;
                } else {
                    // Skip this because too few valid slots have been processed so far.
                    // There are 'many ref newest' accounts in this slot. They must be packed into slots that are >= the current slot value.
                    // We require `min_resulting_packed_slots` target slots. If we have not encountered enough slots already without `many ref newest` accounts, then keep trying.
                    // On the next pass, THIS slot will be older relative to newly ancient slot #s, so those newly ancient slots will be higher in this list.
                    self.shrink_ancient_stats
                        .many_ref_slots_skipped
                        .fetch_add(1, Ordering::Relaxed);
                    remove.push(i);
                    continue;
                }
            }
```

**Comment by jeffwashington** [comment]
> But we can do better, I think.

yes, we can do better. I imagined a few slots wouldn't make a difference. I don't necessarily always want to put the multiref ones first. I want to be able to organize by pubkey or other ways. So, to keep it simple and conservative, I just grab the largest amount. But, yes, you are right. I'd like to see that this algorithm works stable with these parameters.

**Review by brooksprumo — APPROVED** [review]
:shipit:

**Pull request merged**